src/soter/openssl/soter_engine.h

```diff
@@ -54,7 +54,7 @@ struct soter_asym_ka_type {
 };
 
 struct soter_sign_ctx_type {
-    EVP_PKEY_CTX* pkey_ctx;
+    EVP_PKEY* pkey;
```

Comment:
> we stop storing EVP_KEY_CTX because if used only once for key generation and initialization and after that we don't use it directly

Yes.

> And now we store EVP_PKEY, which can store private and public keys, that are only used in all further function calls, yes?

Correct.

> and that is main reason, and we stop store more than we need data, correct?

Yes and no. We don't *need to* free EVP_PKEY_CTX but we don't *have to* retain it either.

As you said, all following methods on `soter_sign_ctx_t` do not use EVP_PKEY_CTX. We only need it for key generation, which is performed exactly once on initialization (at least for `soter_sign_ctx_t`). After that we don't need the context around, only the keypair (or the public key alone). It's not possible to, say, generate a new keypair once again, like `soter_asym_ka_alg_t` allows. In that case it could be viable to retain a preconfigured EVP_PKEY_CTX that can be used for key generation again and again.

However, the main motivation for removing EVP_PKEY_CTX is to prevent the flawed notion from spreading further. EVP_PKEY_CTX is not a general-purpose container for EVP_PKEY, but rather a transient helper. It matters how EVP_PKEY_CTX is created, what for it has been initialized, etc. It's much easier to check correctness of EVP_PKEY_CTX handling when it's constrained to one function, not retained in an object with multiple methods that can affect and reconfigure the context.